lib/api/src/ptr.rs

```diff
@@ -151,65 +131,40 @@ impl<T: Copy + ValueType> WasmPtr<T, Array> {
-        let memory_size = memory.size().bytes().0;
-
-        if (self.offset as usize) + (item_size * slice_full_len) > memory_size
-            || self.offset as usize >= memory_size
-            || mem::size_of::<T>() == 0
-        {
-            return None;
-        }
-
-        unsafe {
-            let cell_ptr = align_pointer(
-                memory.view::<u8>().as_ptr().add(self.offset as usize) as usize,
-                mem::align_of::<T>(),
-            ) as *const Cell<T>;
-            let cell_ptrs = &std::slice::from_raw_parts(cell_ptr, slice_full_len)
-                [index as usize..slice_full_len];
-            Some(cell_ptrs)
-        }
-    }
-
-    /// Mutably dereference this `WasmPtr` getting a `&mut [Cell<T>]` allowing for
-    /// direct access to a `&mut [T]`.
-    ///
-    /// # Safety
-    /// - This method does not do any aliasing checks: it's possible to create
-    ///  `&mut T` that point to the same memory. You should ensure that you have
-    ///   exclusive access to Wasm linear memory before calling this method.
-    #[inline]
-    pub unsafe fn deref_mut(
+    pub fn deref<'a>(
         self,
-        memory: &Memory,
+        memory: &'a Memory,
         index: u32,
         length: u32,
-    ) -> Option<&mut [Cell<T>]> {
+    ) -> Option<Vec<WasmCell<'a, T>>> {
         // gets the size of the item in the array with padding added such that
         // for any index, we will always result an aligned memory access
         let item_size = mem::size_of::<T>();
         let slice_full_len = index as usize + length as usize;
         let memory_size = memory.size().bytes().0;
 
-        if (self.offset as usize) + (item_size * slice_full_len) > memory.size().bytes().0
-            || self.offset as usize >= memory_size
-            || mem::size_of::<T>() == 0
+        if (self.offset as usize) + (item_size * slice_full_len) > memory_size
+            || (self.offset as usize) >= memory_size
+            || item_size == 0
         {
             return None;
         }
 
-        let cell_ptr = align_pointer(
-            memory.view::<u8>().as_ptr().add(self.offset as usize) as usize,
-            mem::align_of::<T>(),
-        ) as *mut Cell<T>;
-        let cell_ptrs = &mut std::slice::from_raw_parts_mut(cell_ptr, slice_full_len)
-            [index as usize..slice_full_len];
-        Some(cell_ptrs)
+        Some(
+            (0..(length as usize))
+                .map(|i| unsafe {
+                    let cell_ptr = align_pointer(
+                        memory
+                            .view::<u8>()
+                            .as_ptr()
+                            .add((self.offset as usize + i * item_size))
+                            as usize,
+                        mem::align_of::<T>(),
+                    ) as *const Cell<T>;
+                    WasmCell::new(&*cell_ptr)
+                })
+                .collect::<Vec<_>>(),
```

Comment:
⛳ 
This can be written with a smaller unsafe block and doing less work as:

```Rust
let cell_ptr = align_pointer(memory.view::<u8>().as_ptr().add(self.offset as usize) as usize, mem::align_of::<T>()) as *const Cell<T>;
let cell_ptrs = &std::slice::from_raw_parts(cell_ptr, slice_full_len)
            [index as usize..slice_full_len];
cell_ptrs.iter().map(|ptr| WasmCell::new(unsafe { &*cell_ptr })).collect::<Vec<_>>()
```

Reply:
Done!